SmartAudio library implementation

wapophis opened the pull request:
This is smartaudio protocol implementatation to deal with the vtx directly throught serial port interface.

That's is a wip, is tested with 2.0 spec. The idea behind this PR is continue working with using this PR as a continuous improvement process. 

Help debugging in 1.0 and 2.1 vtxs is welcome

The setup is the following:

* Wiring: Connect the tx of one spare uart of your fc to the sma pin in the vtx. Also connect the gnd from uart to the gnd of the vtx, be carefull to not mistake with the - pole from power.
* Setup the uart port protocol to value 35 to enable sma protocol in this uart, baud setup is in the sma driver settled to 4800 baud as spec says. This is made setting the value into the SERIAL_<<UART_NUMBER>> protocol param. Setup uart options to half-duplex. 
* Enable vtx setup setting to 1 the para VTX_ENABLE, save and reboot.
* Enable sma protocol in vtx setting the VTX_SMA_ENABLE PARAM TO 1, save and reboot.
* Setup your desired params for band,chan and power in the VTX_BAND,VTX_CHAN,VTX_POWER.
* Setup the additional options which affects pitmode, or unlocking features for the vtx.
* Save and reboot.
When the vtx has power on, the sma driver will send to the hw-vtx the commands needed to sincro your vtx params config with the hw-vtx.
Thats all for now.

andyp1per commented:
@wapophis thanks for working on this! The PR will need quite a few changes to be in a mergeable state, here are the high level changes I think are needed - probably not worth getting into the details until we get a bit closer:

- There is no need for the frontend/backend split, this makes the code unnecessarily complex and large - it should be possible to consolidate into a much smaller set of files. Flash usage is a priority for us and at the moment I suspect the flash usage of this library will be large.
- You should remove AP_Vtx_Backend and use libraries/AP_RCTelemetry/AP_VideoTX - there is a lot of duplicated code that can be removed in favour of the existing code (or modify the existing code)
- The betaflight C code for smartaudio_protocol needs to be converted into C++ and merged into the AP_SmartAudio implementation
- The update loop and config should be put in libraries/AP_Vehicle rather than Copter. AP_VideoTX already has an update look so I suspect that this could just call into the SA code.
- Any extra config should be incorporated into AP_VideoTX and there rest re-used from AP_VideoTX
- You should use a define in the smart audio library to enable/disable rather than putting in Copter - have a look at libraries/AP_Camera/AP_RunCam.h for an example
- If I read the code correctly it is attempting to be a state machine - I think this probably makes the code more complicated than it needs to be and could be removed in favour of a thread. betaflight does not have support for threads which is why they have done it this way - but a state machine is also possible and may be ok once some of the other complexity has been removed.

I'm happy to help you with this if this does not make sense.

wapophis commented:
Thanks for the support.
I'll be posting questions in this thread.

wapophis commented:
Working with the AP_Vehicle i see that the params root is defined here and not in the Parameters.cpp. Is correct to move the root param definition to the AP_Vehicle ?

andyp1per commented:
@wapophis - I'm not quite sure what you mean but what you have done is correct - there is no need to modify Parameters.cpp.

Yeah VTX is protocol independent and small which is why it is universally enabled. The only implemented protocol so far is CRSF which does have the appropriate ifdef

wapophis commented:
I'm thinking about the thread strategy you refer, using the The AP_Scheduler system is the way ?

wapophis commented:
Previous commits refers to new locking param to express locking behaviour into the VTX ( I THINKS THIS IS VERY STANDARD) and the pitmodes which can be set in "_options". I do not know the pitmode behaviour by default in no smartaudio devices, but must be "in range" or "out range", so the actual param 0 could pass to be pitmode disabled.

andyp1per commented:
@wapophis - don't use the scheduler in RCTelemetry, that's to fit telemetry into a limited bandwidth link. The way to create a thread is something like:
```
    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_GyroFFT::update_thread, void), "apm_fft", FFT_STACK_SIZE, AP_HAL::Scheduler::PRIORITY_IO, 1)) {
        AP_HAL::panic("Failed to start AP_GyroFFT update thread");
        return false;
    }
```
but like I said this may not be necessary as you can probably massively reduce the complexity of the implementation without this.

Happy for you to propose different semantics for VTX_OPTIONS or create a separate VTX_PITMODE flag if that makes sense. What is the workflow with "in-range" and "out-range"?

wapophis commented:
That's is the snippet from the loop i working at, take an eye and give me your feedback 
`void AP_SmartAudio::loop(){

    // initialise uart (this must be called from within tick b/c the UART begin must be called from the same thread as it is used from)
    _port->begin(AP_SERIALMANAGER_SMARTAUDIO_BAUD, AP_SERIALMANAGER_SMARTAUDIO_BUFSIZE_RX, AP_SERIALMANAGER_SMARTAUDIO_BUFSIZE_TX);

    // initial setup vtx call
    request_settings();

     while (true) {
        // now time to control loop switching
        uint32_t _now=AP_HAL::millis();
        // when one packet is processed in the iteration will be true.
        bool packet_processed=false;
        monkey_testing();
        // setup a queue polling delay of 20 ms.
        hal.scheduler->delay(20);
        // command to process
        packet _current_command;
        // printf(" time-passed-since-last-request:%d %20s %d ",now-last_request_sended_at,"packet-processed:",packet_processed);

        // Proccess response in the next 200 milis from the request are sent.
        if(_now-last_request_sended_at<=200 && is_waiting_response){
            printf("%20s %d and %d ms more","I'M WAITING RESPONSE SINCE",last_request_sended_at,200-(_now-last_request_sended_at));
            // setup sheduler delay to 20 ms again after response processed
            hal.scheduler->delay(20);
            packet_processed=true;
        }

        // when pending request and last request sended is timeout, take another packet to send
        if(!packet_processed && requests_queue.pop(_current_command)){
            send_request(_current_command.frame,_current_command.frame_size);
            _current_command.sended_at=AP_HAL::millis();
            last_request_sended_at=AP_HAL::millis();
            is_waiting_response=true;
            // spec says: The Unify Pro response frame is usually send <100ms after a frame is successfully received from the host MCU
            printf(" delay: %d",100);
            hal.scheduler->delay(100);
        }
     }
}`

I have some doubs about the setup for the right priorities to setup in the thread instance. I'm working with this now, after reading some libs, and the docs, i setup priority at 60 (same as uart in the doc explanation) :

`hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_SmartAudio::loop, void),
                                          "SmartAudio",
                                          1024, AP_HAL::Scheduler::PRIORITY_UART, 60)`

About the settings about pit mode, we can talk about it later, the spec and the betaflight implementation make some asumptions that i want to clarify in order to make you a proposal. 

Thanks in advance.

wapophis commented:
Commit af7abca can be ignored finally the internal data structure.to mantain vtx status as a representation of the vtx response is the smartaudiosettings_s struct that later Will be refactored

andyp1per commented:
You should set the priority to IO_PRIORITY

My doubts about using a thread are that this is only applicable when not armed,  but the thread will be active (and most importantly the stack space) when flying.

wapophis commented:
> My doubts about using a thread are that this is only applicable when not armed, but the thread will be active (and most importantly the stack space) when flying.

That's an important point to be considered, because it will be usefull to setup the vtx power when flying in order to optimize energy waste. Anyway that feature could be discused at future. 
In ardupiltot docs there is no info about the behaviour you refer :O

wapophis commented:
This is the updated status if the overall changes. 
About the stack Dize, are there anyway yo check It? 

>
> * There is no need for the frontend/backend split, this makes the code unnecessarily complex and large - it should be possible to consolidate into a much smaller set of files. Flash usage is a priority for us and at the moment I suspect the flash usage of this library will be large.
Actualized status: Removed

> * You should remove AP_Vtx_Backend and use libraries/AP_RCTelemetry/AP_VideoTX - there is a lot of duplicated code that can be removed in favour of the existing code (or modify the existing code)
Actualized status: Partially integrated.

> * The betaflight C code for smartaudio_protocol needs to be converted into C++ and merged into the AP_SmartAudio implementation

Actualized status: need your support. Any example plz? 

> * The update loop and config should be put in libraries/AP_Vehicle rather than Copter. AP_VideoTX already has an update look so I suspect that this could just call into the SA code.

Actualized status: fixed

> * Any extra config should be incorporated into AP_VideoTX and there rest re-used from AP_VideoTX

Actualized status: in progress

> * You should use a define in the smart audio library to enable/disable rather than putting in Copter - have a look at libraries/AP_Camera/AP_RunCam.h for an example
Actualized status: fixed

> * If I read the code correctly it is attempting to be a state machine - I think this probably makes the code more complicated than it needs to be and could be removed in favour of a thread. betaflight does not have support for threads which is why they have done it this way - but a state machine is also possible and may be ok once some of the other complexity has been removed.

Actualized status: refactored

> * You should remove all the whitespace changes from files unrelated to this

andyp1per commented:
- So for example libraries/AP_BLHeli/ is code that has come from betaflight and been converted
- You can get the stack size by ftp'ing @SYS/threads.txt from mavproxy
- Conversion to use AP_VideoTX - I am going to work on Tramp support, I can probably give an example of this in the process

wapophis commented:
> * Conversion to use AP_VideoTX - I am going to work on Tramp support, I can probably give an example of this in the process

Integration with AP_VideoTX is almost complete, i'm writing the values directly to the singleton when call get_readings method is called, before i fetch the internal SA values from the buffer, and make any transformación needed in the same method. Because argument to get_readings is a pointer it's support múltiple vtx configuration if needed

andyp1per commented:
@wapophis great! Just make sure that your code copes with something else writing these values - so for instance we have Spektrum VTX support which will allow these values to be written by the TX in which case your driver should be able to detect the change and send the appropriate SA commands

wapophis commented:
@andyp1per good morning, thats the sequence i'm thinking to process external updates from AP_VideoTX, what's your opinion? 
![Third party interaction with VTX throught AP_VideoTX](https://user-images.githubusercontent.com/530987/92218629-af7f8b80-ee99-11ea-9c00-3da9ab7f70e4.png)

andyp1per commented:
@wapophis I think this is mostly right. You can avoid polling updates through the cached copies held by AP_VideoTX - this is what CRSF does, it calls have_params_changed() to decide whether an update is necessary and then loops through the settings calling for instance update_power() which returns true if the power needs updating.

wapophis commented:
@andyp1per  questions about the booting process. When vtx boots, we can query hw-vtx to get his settings, then we can store the settings into ap_vtx, or get the settings from the ap_vtx and update the hw-vtx settings to set the values defined by the user. 
What do you think about setting a param (FIRST_BOOT for example) which models this behaviour ?

andyp1per commented:
@wapophis I'm not sure its worth it as its a first time configuration issue. We already have this issue with configuring the UART and making sure the VTX is speaking the right protocol. Better to just doc in the wiki I think rather than adding a new parameter.

wapophis commented:
@andyp1per Some commits are squashed but i'm not sure if the result is the desired. I'd never before used the squash. Hope the resumed view helps the revision.  
I'm going to start the hardware testing if my ftdi adapter works as expected, unfortunatelly only v1 protocol will be tested. 

I'll be waiting your review, all the overall points you specified at pr start has been reviewed. 

The most important changes are relates to threading, updates flow and smartaudio.cpp integration into lib.

I've started the example tests but i couldn't get It to work because a seg fault. 

Thanks in advance

andyp1per commented:
@wapophis doesn't look squashed, you still have 166 commits - you need to squash these into 1 commit per module, each commit message starting with the module name, e.g. AP_SmartAudio: do stuff
Here's a squash tutorial: https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec
You also need to rebase the PR as several files have changed since you modified them,

andyp1per commented:
I pulled your branch and built for a Durandal - the code does not build (I had to remove crc8_dvb_s2_update) and when I had it built and loaded I was unable to connect via mission planner - so something is messed up in the serial handling with this change. You'll need to at least get it to a state that you can get some access to the board.

wapophis commented:
Ok, i'm reviewing the code. Thanks Andy

wapophis commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -36,6 +36,8 @@ const AP_Param::GroupInfo AP_SmartAudio::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("DEFAULTS",     3, AP_SmartAudio::configured_smartaudio_params, setup_defaults, 0),
 
+    AP_GROUPEND
```
This is the problem which is not failing using sitl but fails on real fc.

wapophis commented:
@andyp1per Good Morning Andy, i'm in the testing process of the sma 2.0 spec, which is the vtx that's i have available, which some changes all seems working fine.
There is a lot of debugging info added in my local repo AP_SmartAudio, are you interested in testing the pr with all the debug traces included in codebase? Or prefer to end the debug process, then clean and push ?

Thanks in advance

andyp1per commented:
I'll try it thanks - trye and get the CI builds above clean.

wapophis commented:
Three last commits, make the lib to work. At least using sma 2.0 spec. Tested with chan,band,power and options changes updated from params defined at AP_VTX namespacel.

andyp1per commented:
Ok so I think I have this set correctly. My VTX is SA 2.1. The output looks like this for the most part:
```
LOOP:Updating HW VTX from AP VTX status
                                                                  VTX-SMARTAUDIO ::request_settings()
 LOOP: Checking Responses
 LOOP: Disabling waiting response
 LOOP: Checking Request
  LOOP MAKING REQUEST TO SEND AT 1223360 ms
  REQ-SEND 5 bytes
  REQ-SEND bytes:AA
  REQ-SEND bytes:55
  REQ-SEND bytes:03
  REQ-SEND bytes:00
  REQ-SEND bytes:9F
 -------------->
 AA
 :


  NEXT LOOP DELAYED 100 ms
 LOOP:Updating HW VTX from AP VTX status
 is_waiting_response:1
 Breaking the update ...
 LOOP: Checking Responses
 I'M WAITING RESPONSE SINCE 1223360 and 899 ms more
  WARNING - VTX-SMARTAUDIO  HW: EMPTY WIRE
 LOOP: scheduler delay 1000
```
So I'm not convinced its working for me. It probably would help to cut down the debug to somr of of the more useful messages as its hard to see whether you have success or failure. I'll do a more detail code review now its close.

andyp1per commented on ArduCopter/APM_Config.h:
```diff
@@ -70,3 +70,4 @@
 //#define USERHOOK_SUPERSLOWLOOP userhook_SuperSlowLoop();  // for code to be run at 1hz
 //#define USERHOOK_AUXSWITCH ENABLED                        // for code to handle user aux switches
 //#define USER_PARAMS_ENABLED ENABLED                       // to enable user parameters
+//#define SMARTAUDIO_ENABLED    DISABLED            // disable TBS SmartAudio support
```
remove

wapophis replied:
Removed

andyp1per commented on ArduCopter/Parameters.cpp:
```diff
@@ -215,7 +215,7 @@ const AP_Param::Info Copter::var_info[] = {
     // @Increment: 10
```
Shouldn't be any changes in this file

andyp1per commented on ArduCopter/Parameters.h:
```diff
@@ -188,7 +188,7 @@ class Parameters {
 
         // 97: RSSI
```
Shouldn't be any changes in this file

andyp1per commented on ArduCopter/RC_Channel.cpp:
```diff
@@ -381,7 +381,7 @@ void RC_Channel_Copter::do_aux_function(const aux_func_t ch_option, const AuxSwi
 
```
remove changes

andyp1per commented on ArduCopter/system.cpp:
```diff
@@ -49,7 +49,7 @@ void Copter::init_ardupilot()
 
```
remove changes

andyp1per commented on ArduCopter/wscript:
```diff
@@ -31,6 +31,7 @@ def build(bld):
             'AP_OSD',
             'AC_AutoTune',
             'AP_KDECAN',
+            'AP_SmartAudio',
```
move to cross-vehicle script

andyp1per commented on libraries/AP_Math/crc.cpp:
```diff
@@ -13,17 +13,17 @@
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 /*
-  collection of CRCs. 
```
removr changes in this file

andyp1per commented on libraries/AP_Math/crc.h:
```diff
@@ -13,15 +13,15 @@
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 /*
-  interfaces to ArduPilot collection of CRCs. 
```
remove changes in this file

andyp1per commented on libraries/AP_RCTelemetry/AP_VideoTX.cpp:
```diff
@@ -61,9 +61,16 @@ const AP_Param::GroupInfo AP_VideoTX::var_info[] = {
     // @Description: Video Transmitter Options.
     // @User: Advanced
-    // @Bitmask: 0:Pitmode
+    // @Bitmask: 0:Pitmode disabled,1: PitMode enabled
```
this is wrong. This is a bitmask so value 0 is disabled, value 1 (bit 0) is enabled

wapophis replied:
So in order to map options : // @Bitmask: 0:Disabled,1: PitMode Enabled, 2:Unlocked

wapophis replied:
So in order to map options : // @Bitmask: 0:Disabled,1: PitMode Enabled, 2:Unlocked

wapophis replied:
Fixed, aditional param for locking, init and additional method removed. 
Added at VideoOptions enum new value for locking.

andyp1per commented on libraries/AP_RCTelemetry/AP_VideoTX.cpp:
```diff
@@ -61,9 +61,16 @@ const AP_Param::GroupInfo AP_VideoTX::var_info[] = {
+    // @Description: Video Transmitter Locked status.
+    // @User: Advanced
+    // @Bitmask: 1:Unlocked, 0 Locked
+    AP_GROUPINFO("LOCKING",  7, AP_VideoTX, _locking, 1),
```
Just add this as a bit to VTX_OPTIONS - no need to create another parameter

wapophis replied:
Resolved using options.

andyp1per commented on libraries/AP_SerialManager/AP_SerialManager.h:
```diff
@@ -106,6 +106,12 @@
 #define AP_SERIALMANAGER_MSP_BAUD           115200
 
+// SmartAudio Protocol
+#define AP_SERIALMANAGER_SMARTAUDIO_BAUD            4800
```
These should go in AP_SmartAudio.h

andyp1per commented on libraries/AP_SerialManager/AP_SerialManager.h:
```diff
@@ -151,14 +157,15 @@ class AP_SerialManager {
         SerialProtocol_MSP = 32,
         SerialProtocol_DJI_FPV = 33,
         SerialProtocol_Sagetech = 34,
+        SerialProtocol_SmartAudio=35,
```
Please observe whitespace conventions in our coding standards. Here and elsewhere: https://ardupilot.org/dev/docs/style-guide.html there is a tool you can use to get you started on the new code.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+#include <stdio.h>
+
+#ifndef SMARTAUDIO_ENABLED
+#define SMARTAUDIO_ENABLED !HAL_MINIMIZE_FEATURES && !APM_BUILD_TYPE(APM_BUILD_Replay)
```
You are not allowed vehicle specific defines in a header. Please remove Replay and #define in the cpp file whatever is causing the issue for Replay

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+#include <AP_Param/AP_Param.h>
+#include <stdio.h>
+
+#ifndef SMARTAUDIO_ENABLED
```
This should be HAL_SMARTAUDIO_ENABLED

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+public:
+
+    // request packet to be processed
+    struct packet{
```
Camel case for classes and structures so "Packet"

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    // request packet to be processed
+    struct packet{
+        smartaudioFrame_t frame;
+        uint8_t frame_size=0;
```
Please don't initialize structures or member variables with null defaults - it messes up the bss, rely on the C++ rules for initialization.

wapophis replied:
That's all the rules for initialization that i must consider ? http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-always

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    } PACKED;
+
+    // configured external params
+    struct configured_smartaudio_params{
```
ConfigureSmartAudioParams although this is already scoped inside AP_SmartAudio so could simply be ConfiguredParams

andyp1per replied:
Is there a reason to put the params inside a structure rather than as just member variables?

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+        AP_Int8  enabled;
+        AP_Int8  protocol_version;
+        AP_Int8  setup_defaults;
+    } smartaudio_params;
```
Here and elsewhere the convention is that member variables should start with "_"

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    bool update(bool force);
+
+    // RingBuffer to store outgoing request.
+    ObjectBuffer<packet> requests_queue{SMARTAUDIO_BUFFER_CAPACITY};
```
This will use a lot of flash since this type has not been used before. It would be nice if we could make this common with the runcam packet, but probably not for this PR

wapophis replied:
Why the ringbuffer is a flash eater ? It's seems as a empowered array.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    ObjectBuffer<packet> requests_queue{SMARTAUDIO_BUFFER_CAPACITY};
+
+    // time the last_request is process
+    uint32_t last_request_sended_at=0;
```
Start with "_" don't initialize to 0 as it is the C++ default. Please do this for all member variables

andyp1per replied:
Generally variables representing a time should include the units so:
```
uint32_t _last_request_sent_ms
```

wapophis replied:
Fixed

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    bool is_waiting_response=false;
+
+    // sends a frame over the wire
+    void send_request(smartaudioFrame_t requestFrame,uint8_t size);
```
Formatting here and in many other places (should be a space after the ",")

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    bool parse_frame_response(const uint8_t *buffer);
+
+    // to prevent use of semaphores use an vtx_states_queue with 2 elements, so one buffered reading
+    ObjectBuffer<smartaudioSettings_s> vtx_states_queue{2};
```
Another whole definition of ObjectBuffer. At a minimum consolidate the two types.

andyp1per replied:
Semaphores would be fine here. It's low priority and slow - no need for a lock free ring buffer

wapophis replied:
So the way to go is the same at betaflight which uses an array and adittional utility methods to control r/w access with semaphores implemented. What do you think?

andyp1per replied:
Seems fine. I think the current implementation is adding complexity that it's hard to understand. Is this simply so that the update thread can see the current values? Easy enough to mediate these with a semaphore. Can also use volatile for atomic things.

wapophis replied:
You can check this impl finally in an array without sem at commit 97d27177ebfb7e9b801ca16198a61837ee0f3b32

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    void set_operation_mode(uint8_t mode);
+
+    // enqueue a set frecuency request, specifiying if the setted frequency is for pit mode or not
+    void set_frequency(uint16_t frecuency,bool isPitModeFreq);
```
"frequency"

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+    void set_channel(uint8_t chan);
+
+    // enqueue a get pit mode frequency request
+    void get_pit_mode_frequency();
```
You are not returning anything so the name does not make sense

wapophis replied:
Rename to request_pit_mode_frequency()

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+
+
+static uint16_t _get_power_in_mw_from_dbm(uint8_t power){
+     switch (power) {
```
Too big to be inline

andyp1per replied:
Perhaps we could turn these into a table in AP_VideoTX, I'm sure there will be more

wapophis replied:
Not already in AP_VideoTX, waiting your revision prior to move. ;)

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,205 @@
+}
+
+static uint16_t _get_power_in_dbm_from_mw(uint16_t power){
+     switch (power) {
```
Too big to be inline

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+
+#define AP_VTX_DEV_MODE     // DEV TESTS ACTIVATE
+#ifdef AP_VTX_DEV_MODE
+# define debug(fmt, args...)	hal.console->printf("\r " fmt "\n", ##args)
```
Don't think there is any need for the \r

wapophis replied:
The only way i found to have a readable output using putty is \r i, if not used the lines has the format:
line
      line
            line
                     
Anyway i planned to remove the debug traces and send mavlink messages for final version to notify the important messages.

andyp1per replied:
Don't use putty - use mavproxy.py - much easier

wapophis replied:
I started using mavproxy with the mavlink message api to see in console, but in the first debug sessión messages are printed out of order because of the loop freq and the xcess of debugging info. Or there is any other way to deliver the messages with a good throughtput to mavproxy ?

wapophis replied:
Fixed, I didn't know that the mavpproxy also get the serial output directly to de screen.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+#include <AP_Math/crc.h>
+#include <GCS_MAVLink/GCS.h>
+
+#define AP_VTX_DEV_MODE     // DEV TESTS ACTIVATE
```
Usually would be called SA_DEBUG or something

wapophis replied:
Fixed

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+    // @Description: Enable SmartAudio protocol on VTX
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLED",     1, AP_SmartAudio::configured_smartaudio_params, enabled, 0,AP_PARAM_FLAG_ENABLE),
```
So this should be "ENABLE" not "ENABLED" but I don't think you need it - you should activate this feature based on the presence of "VTX_ENABLE" and a serial port being configured with smart audio "35". Otherwise the user has to set 3 flags to get this going which is too much

wapophis replied:
Fixed i remove it

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+    // @Param: VERSION
+    // @DisplayName: SmartAudio version
+    // @Description: SmartAudio version
+    // @Values: 0: Version 1,1:Version 2, 3: Version 2.1
```
Can we determine this at runtime? Would make for a better user experience

wapophis replied:
Fixed, i remove it

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+    // @Description: VTX will be configured with VTX params defined by default.
+    // @Values: 0: Disabled,1: Enabled
+    // @User: Advanced
+    AP_GROUPINFO("DEFAULTS",     3, AP_SmartAudio::configured_smartaudio_params, setup_defaults, 0),
```
Probably needs a different name, but not sure what

wapophis replied:
What do you think about, vtx_setup_override ? The settings getted from HW_VTX will be settled at VTX params.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+
+};
+/**
+ * Constructor
```
valueless comment

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+}
+
+AP_SmartAudio *AP_SmartAudio::singleton;
+AP_VideoTX *ap_video_tx=AP_VideoTX::get_singleton();
```
Don't so this as a static, use AP::vtx() whenever you need the AP_VideoTX object

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+        
+        debug("SmartAudio init");
+
+    if(smartaudio_params.enabled.get()==0){
```
Ditch this and just use VTX_ENABLED and the check below

wapophis replied:
fixed

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+    _port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_SmartAudio, 0);
+    if(_port!=nullptr){
+
+          if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_SmartAudio::loop, void),
```
Ok, REALLY IMPORTANT:
* You can use a much smaller stack than this. When I ran it the stack usage was: SmartAudio    PRI=118 STACK_LEFT=832
* You have set the priority to PRIORITY_IO+60 = 118. That's *way* too high. I would set this to UART priority probably or just ```AP_HAL::Scheduler::PRIORITY_IO, -1``` which will give you one below IO where more important stuff runs

wapophis replied:
How can i check thats values to have them in account in the development ?

andyp1per replied:
In mavproxy.py do ```ftp get @SYS/threads.txt```

wapophis replied:
Thanks this evening i'll try it

wapophis replied:
Fixed, i reduced the stack size to 256 and setted prio to -1

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,947 @@
+        // setup port options
+        _port->set_stop_bits(2);
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        _port->set_options(AP_HAL::UARTDriver::OPTION_HDPLEX);
```
Usually this is left to the user to configure on SERIALx_OPTIONS. Maybe this is ok. If you are going to do this you probably want to set the baud rate as well. I think you also probably don't want blocking writes

wapophis replied:
As i understand, it's seems a design aspect of the spec, so it's hardcoded to talk correctly with the uart. Anyway if you think the other aproach give us some value. I'll remove this.

andyp1per replied:
I'm really not sure. I guess what happens if someone swaps the TX/RX?

wapophis replied:
Good point, i'll research this way.

wapophis replied:
uart params could be mixed? Example: HALFDUPLEX AND SWAPPING ?

wapophis commented:
It seems that the baud changes between 4800 baud and 4900 baud depending of the vtx temperature. Betaflight has some control for error rates and adjust baud rate automatically. So with the actual implementation you must wait some iteration loops to get a response from the vtx in the serial channel. Once the response is received the update method takes control updating all the AP_VTX params values, but you'll see than the fc retries several times because the baud rate again. 

I think that we must take the same approach than betaflight if there is no other. This task is already in the todo.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,955 @@
+        packet _current_command;
+        
+        // Proccess response in the next 200 milis from the request are sent.
+        debug("LOOP: Checking Responses");
```
Cut down the debug - difficult to review like this. Probably need to think about the sleep interval and also what happens when we are armed.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,955 @@
+        ,0X0F & _current_state.pitmodeInRangeActive<<0
+        );
+
+        operation_mode |= _current_state.pitmodeInRangeActive<<0;
```
Brackets around the bitshift operation

wapophis replied:
Fixed in 8d208cdac75896b4404257600404fb1403fa4e8f

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,955 @@
+    // pull line low
+    _port->write((uint8_t)0);
+    // write request
+    for (int i = 0; i < size; ++i) {
```
uint8_t

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,955 @@
+            _current_vtx_settings.update_flags=_vtx_settings.update_flags;
+
+
+            if( _vtx_settings.update_flags & (1 << 0)){ // freq has changed
```
There should be an enum for these values - might even be one AP_VideoTX already

wapophis replied:
Fixed. 

        // Proposed to be into AP_VideoTX
        enum class HWVtxUpdates {
            OVERALL_UPD = 0X0F,
            FREQ_UPD = 1 << 0,
            CHAN_UPD = 1 << 1,
            POW_UPD  =  1 << 2,
            MODE_UPD =  1 << 3,
            NO_UPD   = 0
        };

wapophis replied:
Fixed in 8d208cdac75896b4404257600404fb1403fa4e8f

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,955 @@
+    packet command;
+    command.frame=request;
+    command.frame_size=frame_size;
+    requests_queue.push_force(command);
```
Do you really want push_force() here?

wapophis replied:
The idea behind the push_force is to permit discard commands in case of floods from third party libraries. In theory this thirds parties, will be responsible to check that the command has been executed correctly calling to get_readings and checking if the expected change has been produced. 

I imagine the user interacting with the OSD, and changing options fast to setup the vtx, the idea is that last request ( so last options selected) where persited into the vtx.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,955 @@
+
+    // version 1
+    if (sma_version==SMARTAUDIO_SPEC_PROTOCOL_v1) {
+        if (power==14) {
```
Put these in a a table?

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,955 @@
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
These need to be converted to AP naming and made part of the class

wapophis replied:
So the smartaudio_protocol.h must be in AP_SmartAudio too, isn't it?

andyp1per replied:
Yes I suggest so

andyp1per commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -186,6 +186,7 @@ class RC_Channel {
         TER_DISABLE =         86, // disable terrain following in CRUISE/FBWB modes
         CROW_SELECT =         87, // select CROW mode for diff spoilers;high disables,mid forces progressive
         SOARING =             88, // three-position switch to set soaring mode
+        SA_PWR_LOW_OR_AUTO =  89, // SmartAudio VTX power toggle: low aux for minimal power, high aux for auto power
```
I didn't see the corresponding changes in RC_Channel.cpp

andyp1per commented:
@wapophis looking much better! Needs quite a bit of tidyup for formatting, naming, debug etc

wapophis commented:
@andyp1per  Most of the findings in the initial review are resolved.
More debug has been performed and some fixs are applied. 
Thanks in advance.

andyp1per commented:
Doesn't compile on Durandal:
```
[408/712] Compiling libraries/AP_TempCalibration/AP_TempCalibration.cpp
../../libraries/AP_SmartAudio/AP_SmartAudio.cpp: In member function 'void AP_SmartAudio::_print_state(AP_SmartAudio::smartaudioSettings_t&)':
../../libraries/AP_SmartAudio/AP_SmartAudio.cpp:140:15: error: the compiler can assume that the address of 'state' will never be NULL [-Werror=address]
     if (&state!=nullptr){
         ~~~~~~^~~~~~~~~
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors
```
Also needs a rebase

andyp1per commented:
Compiles now on Durandal:
```
SmartAudio    PRI= 57 STACK_LEFT=104
```
The stack is a little low, I would increase it by 256 bytes maybe

How do I tell if its working? I see a lot of this:
```
saStat: badCode: 0, badLength: 0, badPre: 0, badCrc: 0, outofScopeResp: 0, receivedPkt: 0, sendedPkt: 10
 autobaud: now going up

 autobaud: 4850

  WARNING - VTX-SMARTAUDIO  HW: EMPTY WIRE
Flight battery 100 percent
  WARNING - VTX-SMARTAUDIO  HW: EMPTY WIRE
                                                                  VTX-SMARTAUDIO ::request_settings()
  WARNING - VTX-SMARTAUDIO  HW: EMPTY WIRE
```
But I can't see any messages actually flowing

wapophis commented:
@andyp1per I'll make the debug output more readable. 

Anyway the output you send me, show me that the response buffer size must be increased with sma 2.1 because of the power tables and the splitted reading is failling. 

I'll push this changes soon.

Thanks a lot

andyp1per commented:
```
SA:                                              READ RESPONSE incoming_bytes_count: 17
 READ RESPONSE response_in_bytes:00
SA:  READ RESPONSE byte discard:00
 READ RESPONSE response_in_bytes:AA
SA:  READ RESPONSE byte -> response_buffer added :AA
 READ RESPONSE response_in_bytes:55
SA:  READ RESPONSE byte -> response_buffer added :55
 READ RESPONSE response_in_bytes:11
SA:  READ RESPONSE byte -> response_buffer added :11
 READ RESPONSE response_in_bytes:0D
SA:  READ RESPONSE byte -> response_buffer added :0D
 READ RESPONSE response_in_bytes:19
SA:  READ RESPONSE byte -> response_buffer added :19
 READ RESPONSE response_in_bytes:00
SA:  READ RESPONSE byte -> response_buffer added :00
 READ RESPONSE response_in_bytes:10
SA:  READ RESPONSE byte -> response_buffer added :10
 READ RESPONSE response_in_bytes:16
SA:  READ RESPONSE byte -> response_buffer added :16
 READ RESPONSE response_in_bytes:F8
SA:  READ RESPONSE byte -> response_buffer added :F8
 READ RESPONSE response_in_bytes:0E
SA:  READ RESPONSE byte -> response_buffer added :0E
 READ RESPONSE response_in_bytes:04
SA:  READ RESPONSE byte -> response_buffer added :04
 READ RESPONSE response_in_bytes:00
SA:  READ RESPONSE byte -> response_buffer added :00
 READ RESPONSE response_in_bytes:0E
SA:  READ RESPONSE byte -> response_buffer added :0E
 READ RESPONSE response_in_bytes:14
SA:  READ RESPONSE byte -> response_buffer added :14
 READ RESPONSE response_in_bytes:1A
SA:  READ RESPONSE byte -> response_buffer added :1A
 READ RESPONSE response_in_bytes:1D
SA:  READ RESPONSE byte -> response_buffer added :1D
SA: RESPONSE SEEMS NOT TO BE COMPLETE WAITING IN OTHER LOOP CYCLE ?
SA:                                              READ RESPONSE incoming_bytes_count: 2
 READ RESPONSE response_in_bytes:FD
SA:  READ RESPONSE byte discard:FD
 READ RESPONSE response_in_bytes:00
SA:  READ RESPONSE byte discard:00
SA: INPUT FILTERED FROM UART IS SHORTER THAN EXPLICIT RESPONSE LENGTH
```

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,881 @@
+        // Proccess response in the next 200 milis from the request are sent.
+        if (now-_last_request_sended_at_ms>100 && now-_last_request_sended_at_ms<=1000 && _is_waiting_response){
+            // allocate response buffer
+            uint8_t _response_buffer[AP_SMARTAUDIO_UART_BUFSIZE_RX];
```
Have a look at how RCProtocol does this. I am not against allocating such a small buffer on the stack, but you should cope with not getting all the data at once. 
In RCProtocol we read bytes into a buffer until we have a header, if we don't have enough bytes we loop and sleep. You should do something similar instead of assuming that you have a complete packet in the UART.

wapophis replied:
Fine, i'll check the actual approach which has in account the splitted responses, but seems that i have a bug on it. 
I'll research the sleep approach.

wapophis replied:
After some rework is working with splitted responses, max uart buffer size is 16 bytes, max buffer retained 32 bytes.

andyp1per commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,881 @@
+ **/
+void AP_SmartAudio::read_response(uint8_t *response_buffer, uint8_t inline_buffer_length)
+{
+    int16_t incoming_bytes_count = _port->available();
```
Are you expecting a response here? If so you should keep reading until you get something or timeout - or make the function re-entrant and do this from the thread loop. At the moment you simply give up if there are not enough bytes.

wapophis replied:
Yes this method makes the response reading. Here the splitted response must be captured and completed. I'll check the impl to acomplish well the task.

wapophis replied:
Fixed in db8526f24036d58d4f88554c9750ae65d05869a1

andyp1per commented:
I can't push to your branch - but I've done a PR to tidy up the debug a little so that I can see what is going on. I think perhaps you are not reading a complete packet and that is causing the issues - see my comments earlier.

andyp1per commented:
Here are a couple of examples:

https://github.com/ArduPilot/ardupilot/blob/11b3c622ddab875bbaa8fc3eb74458d19467ed0d/libraries/AP_RCProtocol/AP_RCProtocol.cpp#L258
https://github.com/ArduPilot/ardupilot/blob/11b3c622ddab875bbaa8fc3eb74458d19467ed0d/libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp#L145
Notice how the timeout / bytes available / incomplete message are incorporated into the loop

wapophis commented:
@andyp1per the splitted reading from response buffer are now working. I'd made some test in 4 bytes packets and after some little rework on the read_response is working now. (fixed on db8526f24036d58d4f88554c9750ae65d05869a1)

I'd detected a problem with the timings in the main loop, which makes laggy the syncing changes with vtx, now it's fixed too, performing well in one loop usually.  (fixed on 79ec8d416a9af513b69106dc4c40507d83f0ebe8)

Finally the Ap_SmartAudio, is updating the AP_videoTx values when the hot_deploy param is enabled, syncing params persited in the eprom with default backend params. making the syncing to work without needing to reboot de FC.

Other things to reply:

-  PushForce: The strategy is to discard changes that were repeated. For example: if user is making channel selection fast, the vtx is only interested in the final one, using this approach retains the last five changes in queue. I can add more logic to the input into the queue, discarding events which are repeated and keeping the ones which are diferent type of commands. I think that this feature will be needed, but first i'll consolidate the protocol implementation without features which are not needed (like hot deploy which is basically for debugging)

andyp1per commented:
With the latest I eventually get this, but it takes a long time - feels like something is not quite right
```
SA: send_request(): 0xAA 0x55 0x03 0x00 0x9F
SA: Ap_SmartAudio: byte discard:00
SA: read_response(): 0x55
SA: INPUT FILTERED FROM UART IS SHORTER THAN EXPLICIT RESPONSE LENGTH
SA: request_settings()
SA: send_request(): 0xAA 0x55 0x03 0x00 0x9F
SA: read_response(): 0xAA 0x55 0x11 0x0D 0x19 0x00 0x10 0x16 0xF8 0x0E 0x04 0x00 0x0E 0x14 0x1A 0x1D 0xFD 0x00
SA:                                                     selecting update path using: 0APM: AP_SmartAudio: Operation REQUEST SETTINGS VERSION 2.1 completed since 102 milliseconds
F
SA:                   parse_frame_response:: freq update from band and chan settings
SA: UPDATE AP_VTX->HW_VTX: CHAN
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x07 0x01 0x22 0x63
APM: RESPONSE RETRY:0
SA: send_request(): 0xAA 0x55 0x05 0x01 0x0E 0xC3
SA: send_request(): 0xAA 0x55 0x0B 0x01 0x00 0x2D
SA:
AP_SmartAudio: Printing Status debug info [58280,3067]
SA: AP_SmartAudio: Hardware vtx is initialized, has no pending updates and is waiting response with hot deploy ON
SA: AP_SmartAudio STATS ERRORS [Unknow Responses Types count: 0, Bad lenght Responses count: 0,Bad Headers Responses count: 0, Bad Crc response count: 0, Unparseable Responses count: 414 ]
SA: AP_SmartAudio STATS AUTO ADJUST BAUD STATS [Responses count: 1, Request counter: 9, Baud Rate: 4980, Baud Rate direcction:Down ]
SA: AP_SmartAudio STATS COMMANDS  [ FAILED READS: 414, REQ_SETTINGS: 46, OTHER REQUESTS:3 ]
SA: UPDATE AP_VTX->HW_VTX: CHAN
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x07 0x01 0x22 0x63
SA: send_request(): 0xAA 0x55 0x05 0x01 0x0E 0xC3
SA: send_request(): 0xAA 0x55 0x0B 0x01 0x00 0x2D
```

wapophis commented:
Thanks Andy i take a look

andyp1per commented:
I've spent a fair bit of time on this today. Fixed a few bugs and restructured a few things that were wrong - not quite working correctly, but closer I think. Will put up a PR shortly.

wapophis commented:
Fine, thanks

andyp1per commented:
This is the PR https://github.com/wapophis/ardupilot/pull/4

andyp1per commented:
With my change I now get this:
```
SA: send_request(): 0xAA 0x55 0x03 0x00 0x9F
SA: read_response(): 0x1A 0x1D 0x1C
SA: selecting update path using: 0F
SA:                   parse_frame_response:: freq update from band and chan settings
SA: UPDATE AP_VTX->HW_VTX: Ch: 2, Band: 4
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x07 0x01 0x22 0x63
APM: AP_SmartAudio: Operation REQUEST SETTINGS VERSION 2.1 completed since 160 milliseconds
APM: SmartAudio initialized, SA 2.1, Band 3, Ch 1
APM: RESPONSE RETRY:171
SA: UPDATE AP_VTX->HW_VTX: Ch: 2, Band: 4
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x05 0x01 0x0E 0xC3
SA: UPDATE AP_VTX->HW_VTX: Ch: 2, Band: 4
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x05 0x01 0x0E 0xC3
SA: UPDATE AP_VTX->HW_VTX: Ch: 2, Band: 4
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x05 0x01 0x0E 0xC3
SA: UPDATE AP_VTX->HW_VTX: Ch: 2, Band: 4
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x05 0x01 0x0E 0xC3
SA: UPDATE AP_VTX->HW_VTX: Ch: 2, Band: 4
SA: AP_SmartAudio: Setting channel to 34
SA: UPDATE AP_VTX->HW_VTX: OPTIONS
SA: set_operation_mode(): Changing operation mode to 00
SA: send_request(): 0xAA 0x55 0x05 0x01 0x0E 0xC3
SA:
AP_SmartAudio: Printing Status debug info [91893,3003]
```
Fairly continuously after it finally communicates. The device also seems to switch into 800mw mode - which is not what is configured, so something not correct as yet.

wapophis commented:
I'm checking the pr, one thing that it's seems strange is that the first reading from response buffer seems partial and i dont see the complete response from vtx, and subsequent requests has not been responsed. Other thing i notice is that the pulldown for the requests has been removed. Anyway I'll try with this pr into my FC with sma 2.0 and check if it's already working.

andyp1per commented:
I think you should put the pulldown back in - I just took it out so I could see the trace properly on the oscilloscope

andyp1per commented:
There are still some timing things to work out. I suspect the low priority of the thread is contributing to this.

andyp1per commented:
There is definitely something wrong with the pitmode and power settings. I think I have a fix for the pitmode stuff

andyp1per commented:
Part of the issue is I have totally toasted the lipo I was testing with, will find another ...

andyp1per commented:
Ok, you cannot lock/unlock the VTX via SA it seems - at least not on my Evo, you can only query the setting.

andyp1per commented:
@wapophis I have updated my PR. There are still a fair few bugs in the logic handling of this PR:

- AP_VideoTX already know the power level in dbm, we should use that instead of re-calculating. I have a fixed a couple of instances of this but it needs a bit more cleanup - in particular the logic for SA 2.1 was wrong
- The level 0 byte definitely confuses my Evo and it seems to work perfectly well without it (much better in fact)
- The SA autobaud needs to quickly cycle through the baud rates until it finds something that works and then stay there until a response is missed. 100 ticks is too much I think - you could probably change the baud every 100ms and it would be fine
- The syncing between AP_VideoTX and SA is not right and I think could be cleaned up a fair bit - it is still trying to set settings that do not need to be changed
- Please, please run the code through the AP formatter : https://ardupilot.org/dev/docs/style-guide.html and adhere to this, it makes it much easier to review
- Also please note that my switch of the SA structures to C++ style revealed a fair few initialization bugs - so wherever possible we need to stick to C++ style

I'll do a more detailed review now that it's roughly working for me, out of time for actual testing for now I am afraid

andyp1per commented:
Hey @wapophis what progress here?

wapophis commented:
> Hey @wapophis what progress here?

Yes Andy because of working presure i had not too much time to dev. Next weekend i'll push to the pr. During this dev impass i have been flying whithout any collateral impact.

andyp1per commented:
Hi @wapophis any movement on this? I'll clone your PR and work on it if you are not going to get to it

skorokithakis commented:
This would be fantastic to merge, currently SmartAudio support is the only thing keeping me from using AP, as the VTX can get very hot if it's always set to the highest power, especially testing on the bench.

andyp1per commented:
Ok, I have started work on this

andyp1per commented:
This has been subsumed by #16464 which is now merged
